### ISSUE TITLE
Conflicting tag handling and unused LANG2BYTE in Dia-TTS fine-tuning code

### ISSUE THREAD

**Issue by manhcuong17072002** [TuananhCR/Dia-Finetuning-Vietnamese#7]
Hello,  

I want to fine-tune Dia-TTS, but I noticed that your code has several logical conflicts:  

- In `dataset.py`, the tag used is `[vi]`.  
- In `finetune.py`, it is lowercase `[kienthucquansu]`.  
- During inference, it becomes `[KienThucQuanSu]`.  

Especially, the variable `LANG2BYTE` is not used, and if it is used, it will conflict with other byte characters such as `a, b, c, d`. This may cause errors or misreading due to overlapping byte encodings, instead of using ASCII control characters like in the original code.  

So I think this might not be the complete code you actually used for training.  
Could you please publish the full training code?  

Thanks.

**Comment by TuananhCR** [comment]
1. My baseline training setup:
For voice-specific fine-tuning, these settings were the most stable:
	•	Batch size: usually can’t go too high — I stick with per-GPU batch = 1 ~ 2
→ use grad_accum_steps = 8 ~ 16 to simulate a bigger effective batch (8–16)
	•	Learning rate: 1e-4 with cosine schedule + small warm-up (about 1–2 k steps)
	•	Epochs: around 3–5 (full pass) or roughly 50 k–200 k steps depending on your dataset
	•	Precision: bf16
	•	Grad clip: 1.0
	•	Optimizer: AdamW (default betas are fine)
I trained on a single A6000 (48 GB) and total time per voice dataset was roughly 6–8 hours for ~50 k steps.

2. Batch > 2 → KV-cache shape mismatch

That expanded size of tensor (2) must match existing size (4) error is from the attention cache.
The model uses GQA with 16 query heads / 4 KV heads.

What happens is the cache gets created with [4, 16, T, 128] instead of [batch, 16, T, 128], so it only works when batch accidentally = 4.

Quick fix: inside layers.py → prefill_kv(), make sure the cache is created with the current batch size as the first dimension. For example:
`B, _, T, Hd = k.shape
if self.k is None or self.k.shape[0] != B:
    self.k = k.new_zeros((B, kv_heads, T_capacity, Hd))`
After that, you should be able to raise the batch or just keep small batch + grad accumulation.

3. Yes, that warning comes from the dataset. It means your audio codes include values > 1023, but Dia’s DAC tokenizer only allows 0–1023 (plus 1024–1026 reserved for EOS/PAD/BOS).

Usually this happens when:
	•	the audio was encoded with a different DAC model (wrong codebook size), or
	•	some preprocessing didn’t clip properly.

To check: `codes.min(), codes.max()`
If you see anything > 1023 → you need to re-encode those wavs with the same DAC that Dia uses (44.1 kHz, 9 channels, 1536 frames each). Also make sure your dataset shape is [T, 9] and values only in [0, 1023].

**Comment by manhcuong17072002** [comment]
@TuananhCR  Hello,

I’ve reviewed your model and also encountered the same warning during inference:

```
Warning: Clamping 54 indices outside range [0, 1023] to 0.
E1:  49%|██████████████████████████████████████████▌                                           | 1972/3988 [1:11:23<1:19:16,  2.36s/it, loss=3.5705, VRAM (GB)=9.70/31.51]
Warning: Clamping 15 indices outside range [0, 1023] to 0.
E1:  50%|██████████████████████████████████████████                                           | 1975/3988 [1:11:28<1:03:06,  1.88s/it, loss=3.5254, VRAM (GB)=15.70/31.52]
```

[Segment 1] [HocEnglishOnline] Hello, I’m the AI assistant.

```
Warning: Clamping 12 indices outside range [0, 1023] to 0.
🟩 Total tokens generated: 176
```

[Segment 2] [HocEnglishOnline] Today I’ll help you create a natural-sounding voice.

```
Warning: Clamping 15 indices outside range [0, 1023] to 0.
🟩 Total tokens generated: 206
```

So I think this issue isn’t caused by the data. Do you know if there’s any way to completely fix this problem?

**Comment by TuananhCR** [comment]
It's not really a problem. You can just ignore it then. I did use it to print out several issues before. So you dont need to care about that much.

**Comment by manhcuong17072002** [comment]
I’ve fixed the issue:
In the `model.py` file, change

```python
logits_CxV[:, 1025:] = -torch.inf
```

to

```python
logits_CxV[:, 1025:] = -torch.inf
logits_CxV[1:, 1024:] = -torch.inf
```

However, I’m not sure whether this affects the results or not, but the warning no longer appears. Thanks

**Comment by manhcuong17072002** [comment]
@TuananhCR  **Hello,**

I’m currently testing multi-speaker training for our internal voice dataset, but I’ve encountered the following issues:

* During the first training run (fine-tuned from your checkpoints), I get **random voices** even though the speaker is specified — each inference produces a different voice.
  For example:
  `"[Phương Anh]Hôm nay mình có meeting lúc 10 giờ sáng, chắc phải dậy sớm hơn một chút."`
  → The output voice is completely different from the expected speaker.

* In subsequent training runs, the output becomes **consistent (fixed voice)**, but it’s still not the original trained speaker’s voice.
  For example:
  `"[Phương Anh]Hôm nay mình có meeting lúc 10 giờ sáng, chắc phải dậy sớm hơn một chút."`
  → The generated voice is **male**, even though the original was female.

I have checked the data structure in the Parquet file and found **no mismatch between audio and speaker**.
The data structure in the Parquet file is:

```json
{
  "audio": {
    "array": [],
    "sampling_rate": 44100
  },
  "text": "str",
  "speaker": "str",
  "sample_rate": 44100
}
```

* **Training text format:** `[speaker]text`
* Each speaker has approximately 8,000 samples, equivalent to about 14 hours of audio per person.
* For encoding the `[speaker]` tag, I tried two approaches:

  1. Convert to **controllable byte format** like `b'\x01', b'\x02', ...` as in the original author’s repository.
  2. Convert to **UTF-8 encoding**, for example:
     `'[Phương Anh]'.encode("utf-8") → b'[Ph\xc6\xb0\xc6\xa1ng Anh]'`.

However, both methods result only in **correct pronunciation**, not the **correct speaker voice**.
Could you please help me resolve this issue?

**Comment by manhcuong17072002** [comment]
Thank you for your response.
I found the issue in the dataloader setup — in the finetune repo, `shuffle` was not set to `True`. Meanwhile, my dataset is clustered by speaker, meaning that samples from the same speaker are grouped together consecutively. This caused an issue where the model failed to properly distinguish or respond to the `[speaker tag]` in the input text.

I’ve successfully completed the training now, but I still need to adjust the tags to avoid strange sounds appearing at the beginning of the generated audio (possibly because the tag overlaps with a readable character — that’s my assumption).

Thank you for your help and support.

Thanks again.
